gk/fib.c

```diff
@@ -1533,3 +1534,295 @@ del_fib_entry(const char *ip_prefix, struct gk_config *gk_conf)
+
+	return 0;
+}
+
+static int
+fillup_gateway_fib_dump_entry(
+	struct gk_fib_dump_entry *dentry, struct gk_fib *fib)
+{
+	int ret;
+	struct gk_fib_ether_entry *eth_tbl = rte_calloc(
+		NULL, 1, sizeof(struct gk_fib_ether_entry), 0);
+	if (eth_tbl == NULL)
+		return -1;
+
+	ret = fillup_gk_fib_ether_entry(fib->u.gateway.eth_cache, eth_tbl);
+	if (ret < 0) {
+		rte_free(eth_tbl);
+		return -1;
+	}
+
+	dentry->num_ether_entries = 1;
+	dentry->ether_tbl = eth_tbl;
+
+	return 0;
+}
+
+static int
+fillup_neighbor_fib_dump_entry(uint16_t proto,
+	struct gk_fib_dump_entry *dentry, struct gk_fib *fib)
+{
+	int num_entries = 0;
+	uint32_t next = 0;
+	int32_t index;
+	const void *key;
+	void *data;
+	struct neighbor_hash_table *neigh_ht;
+	struct gk_fib_ether_entry *eth_tbl;
+	struct gk_fib_ether_entry *entry;
+
+	if (proto == ETHER_TYPE_IPv4)
+		neigh_ht = &fib->u.neigh;
+	else
+		neigh_ht = &fib->u.neigh6;
+
+	eth_tbl = rte_calloc(NULL, neigh_ht->tbl_size,
+		sizeof(struct gk_fib_ether_entry), 0);
+	if (eth_tbl == NULL)
+		return -1;
+
+	index = rte_hash_iterate(neigh_ht->hash_table,
+		(void *)&key, &data, &next);
+	while (index >= 0) {
+		int ret;
+		struct ether_cache *eth_cache = &neigh_ht->cache_tbl[index];
+		entry = &eth_tbl[num_entries++];
+
+		ret = fillup_gk_fib_ether_entry(eth_cache, entry);
+		if (ret < 0) {
+			rte_free(eth_tbl);
+			return -1;
+		}
+
+		index = rte_hash_iterate(neigh_ht->hash_table,
+			(void *)&key, &data, &next);
+	}
+
+	dentry->num_ether_entries = num_entries;
+	dentry->ether_tbl = eth_tbl;
+
+	return 0;
+}
+
+static int
+fillup_gk_fib_dump_entry(uint16_t proto,
+	struct gk_fib_dump_entry *dentry, struct gk_fib *fib)
+{
+	int ret = -1;
+	dentry->action = fib->action;
+	switch (fib->action) {
+	case GK_FWD_GRANTOR:
+		ret = fillup_grantor_fib_dump_entry(dentry, fib);
+		break;
+
+	case GK_FWD_GATEWAY_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_GATEWAY_BACK_NET:
+		ret = fillup_gateway_fib_dump_entry(dentry, fib);
+		break;
+
+	case GK_FWD_NEIGHBOR_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_NEIGHBOR_BACK_NET:
+		ret = fillup_neighbor_fib_dump_entry(
+			proto, dentry, fib);
+		break;
+
+	case GK_DROP:
+		ret = 0;
+		break;
+
+	default:
+		rte_panic("Invalid FIB action (%u) at %s with lcore %u\n",
+			fib->action, __func__, rte_lcore_id());
+		break;
+	}
+
+	return ret;
+}
+
+struct gk_fib_dump_entry *
+list_fib_entries(struct gk_config *gk_conf, uint32_t *num_entries)
+{
+	int ret;
+	uint8_t depth = 0;
+	uint32_t next = 0;
+	int index;
+	char ip_addr[INET6_ADDRSTRLEN];
+	struct gk_fib *fib;
+	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
+	unsigned num_entries_max = gk_conf->max_num_ipv4_rules +
+		gk_conf->max_num_ipv6_rules;
+	struct gk_fib_dump_entry *tbl;
+	const struct rte_lpm_rule *re4;
+	const struct rte_lpm6_rule *re6;
+
+	if (num_entries == NULL)
+		return NULL;
+	*num_entries = 0;
+
+	tbl = rte_calloc("dump_entry", num_entries_max,
+		sizeof(struct gk_fib_dump_entry), 0);
+	if (tbl == NULL)
+		return NULL;
```

Comment:
I think we should have an error message here saying that the memory allocation was the reason for the failure.